neonvm-runner/cmd/main.go

```diff
@@ -669,6 +673,85 @@ func main() {
 	}
 }
 
+type NetworkMonitoringMetrics struct {
+	IngressBytes, EgressBytes       prometheus.Counter
+	Errors                          *prometheus.CounterVec
+	IngressBytesRaw, EgressBytesRaw uint64 // Absolute values to calc increments for Counters
+}
+
+func NewMonitoringMetrics(reg *prometheus.Registry) *NetworkMonitoringMetrics {
+	m := &NetworkMonitoringMetrics{
+		IngressBytes: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "runner_vm_ingress_bytes",
+				Help: "Number of bytes received by the VM from the open internet",
+			},
+		),
+		EgressBytes: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "runner_vm_egress_bytes",
+				Help: "Number of bytes sent by the VM to the open internet",
+			},
+		),
+		IngressBytesRaw: 0,
+		EgressBytesRaw:  0,
+		Errors: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "runner_vm_network_fetch_errors_total",
+				Help: "Number of errors while fetching network monitoring data",
+			},
+			[]string{"cause"},
+		),
+	}
+	reg.MustRegister(m.IngressBytes)
+	reg.MustRegister(m.EgressBytes)
+	reg.MustRegister(m.Errors)
+	return m
+}
+
+func getNetworkBytesCounter(iptables *iptables.IPTables, chain string) (uint64, error) {
+	cnt := uint64(0)
+	rules, err := iptables.Stats("filter", chain)
+	if err != nil {
+		return cnt, err
+	}
+	for _, rawStat := range rules {
+		stat, err := iptables.ParseStat(rawStat)
+		if err != nil {
+			return cnt, err
+		}
+		if stat.Protocol == "6" { // tcp
+			cnt += stat.Bytes
+		}
+	}
+	return cnt, nil
+}
+
+func (m *NetworkMonitoringMetrics) increment() {
+	// Rules configured at github.com/neondatabase/cloud/blob/main/compute-init/compute-init.sh#L98
+	iptables, err := iptables.New()
+	if err != nil {
+		m.Errors.WithLabelValues(util.RootError(err).Error()).Inc()
+		return
+	}
+
+	if ingress, err := getNetworkBytesCounter(iptables, "INPUT"); err != nil {
+		m.Errors.WithLabelValues(util.RootError(err).Error()).Inc()
+		return
+	} else {
+		m.IngressBytes.Add(float64(ingress - m.IngressBytesRaw))
+		m.IngressBytesRaw = ingress
+	}
```

Comment:
Here, I would recommend rewriting this to put `ingress, err :=` on its own line, and then move the contents of the `else { ... }` block in the main body of the function.

That way the `if err != nil { ... }` is easier to visually skip over to focus reading on just the happy-path.

\+ same comment for the ones below

Reply:
My personal preference is this style as it's more terse. If we don't have this in the style guide I'd prefer to leave this as is